metricflow/query/group_by_item/candidate_push_down/group_by_item_candidate.py

```diff
@@ -0,0 +1,97 @@
+from __future__ import annotations
+
+import itertools
+from dataclasses import dataclass
+from typing import Iterable, Tuple
+
+from typing_extensions import override
+
+from metricflow.query.group_by_item.path_prefixable import PathPrefixable
+from metricflow.query.group_by_item.resolution_dag.resolution_nodes.base_node import GroupByItemResolutionNode
+from metricflow.query.group_by_item.resolution_path import MetricFlowQueryResolutionPath
+from metricflow.specs.patterns.spec_pattern import SpecPattern
+from metricflow.specs.specs import InstanceSpecSet, LinkableInstanceSpec, LinkableSpecSet
+
+
+@dataclass(frozen=True)
+class GroupByItemCandidateSet(PathPrefixable):
+    """The set of candidate specs that could match a given spec pattern.
+
+    This candidate set is refined as it is passed from the root node (representing a measure) to the leaf node
+    (representing a query) in the resolution DAG.
+    """
+
+    specs: Tuple[LinkableInstanceSpec, ...]
+    # The path from the leaf node to the measure node where the specs originated. There can be multiple paths as
+    # when candidate sets are merged, there can be multiple measures were a spec could have come from.
+    measure_paths: Tuple[MetricFlowQueryResolutionPath, ...]
+    # The path from the leaf node to the node where this set was generated.
+    path_from_leaf_node: MetricFlowQueryResolutionPath
```

Comment:
Two things:

1. I think these inline comments are better placed in the class docblock for a host of reasons
2. It would be nice to have an indication of what these last two properties are used for - if it's error reporting, that warrants a different level of scrutiny (and possibly a separate container class) than if these interact with the valid spec set we're generating in these push down operations